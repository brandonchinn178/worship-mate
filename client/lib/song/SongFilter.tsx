import * as _ from 'lodash'
import styled, { css } from 'styled-components'

import { color, font } from '~/theme'

import {
  ActiveFilters,
  AvailableFilterOptions,
  AvailableFilters,
  FilterHandler,
  FilterNames,
} from './filters'

type SongFilterContext = {
  activeFilters: ActiveFilters
  filterHandler: FilterHandler
}

type SongFilterProps = SongFilterContext & {
  availableFilters: AvailableFilters
}

export function SongFilter({ availableFilters, ...props }: SongFilterProps) {
  return (
    <FilterBox>
      {_.map(availableFilters, (options, name) => (
        <SongFilterSection
          key={name}
          filterName={name as FilterNames}
          filterOptions={options}
          {...props}
        />
      ))}
    </FilterBox>
  )
}

type SongFilterSectionProps<Name extends FilterNames> = SongFilterContext & {
  filterName: Name
  filterOptions: AvailableFilterOptions<Name>
}

function SongFilterSection<Name extends FilterNames>({
  filterName,
  filterOptions,
  activeFilters,
  filterHandler,
}: SongFilterSectionProps<Name>) {
  return (
        <FilterCategory>
          <FilterLabel>{_.startCase(filterName)}</FilterLabel>
          <FilterOptionGrid>
            {filterOptions.map(({ value, valueDisplay, count }) => {
              const isActive =
                filterOptions.length === 1 || activeFilters[filterName] === value
              return (
                <FilterOption
                  active={isActive}
                  key={valueDisplay}
                  onClick={() => {
                    if (isActive) {
                      filterHandler.removeFilter(filterName)
                    } else {
                      filterHandler.addFilter(filterName, value)
                    }
                  }}
                >
                  {valueDisplay} ({count})
                </FilterOption>
              )
            })}
          </FilterOptionGrid>
        </FilterCategory>
  )
}

const FilterBox = styled.div`
  padding: 20px;

  border: 1px solid ${color('black')};
  box-shadow: 0 0 4px ${color('lightGray')};
`

const FilterCategory = styled.div`
  &:not(:last-child) {
    margin-bottom: 1rem;
  }
`

const FilterLabel = styled.h1`
  ${font('label')}
  color: ${color('darkTeal')};
  font-size: 1.5em;
`

const FilterOptionGrid = styled.div`
  display: flex;
  flex-wrap: wrap;
`

const FilterOption = styled.div<{ active: boolean }>`
  margin: 2px;
  padding: 5px 10px;

  font-weight: bold;

  border: 2px solid ${color('black')};
  border-radius: 3px;
  cursor: pointer;

  ${(p) =>
    p.active &&
    css`
      color: ${color('white')};
      background: ${color('darkTeal')};
    `}
`
