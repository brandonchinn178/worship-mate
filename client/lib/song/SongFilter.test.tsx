import { fireEvent, wait } from '@testing-library/react'

import { renderUI } from '~jest-utils'

import SongFilter from './SongFilter'

const testFilters = [
  {
    key: 'recommendedKey',
    options: [
      {
        name: 'C',
        count: 1,
      },
      {
        name: 'G',
        count: 2,
      },
      {
        name: 'E',
        count: 3,
      },
    ],
  },
  {
    key: 'timeSignature',
    options: [
      {
        name: '4/4',
        count: 6,
      },
    ],
  },
]

it('renders filters', () => {
  const { getByText } = renderUI(
    <SongFilter
      filters={testFilters}
      activeFilters={{}}
      addFilter={jest.fn()}
      removeFilter={jest.fn()}
    />,
  )

  expect(getByText('Recommended Key')).toBeVisible()
  expect(getByText('C (1)')).toBeVisible()
  expect(getByText('G (2)')).toBeVisible()
  expect(getByText('E (3)')).toBeVisible()
  expect(getByText('Time Signature')).toBeVisible()
  expect(getByText('4/4 (6)')).toBeVisible()
})

it('can add a filter', async () => {
  const mockAddFilter = jest.fn()

  const { getByText } = renderUI(
    <SongFilter
      filters={testFilters}
      activeFilters={{}}
      addFilter={mockAddFilter}
      removeFilter={jest.fn()}
    />,
  )

  fireEvent.click(getByText('G (2)'))

  await wait(() => expect(mockAddFilter).toHaveBeenCalled())
  expect(mockAddFilter.mock.calls[0]).toEqual(['recommendedKey', 'G'])
})

it('can change a filter', async () => {
  const mockAddFilter = jest.fn()

  const { getByText } = renderUI(
    <SongFilter
      filters={testFilters}
      activeFilters={{ recommendedKey: 'G' }}
      addFilter={mockAddFilter}
      removeFilter={jest.fn()}
    />,
  )

  fireEvent.click(getByText('E (3)'))

  await wait(() => expect(mockAddFilter).toHaveBeenCalled())
  expect(mockAddFilter.mock.calls[0]).toEqual(['recommendedKey', 'E'])
})

it('can remove a filter', async () => {
  const mockRemoveFilter = jest.fn()

  const { getByText } = renderUI(
    <SongFilter
      filters={testFilters}
      activeFilters={{ recommendedKey: 'G' }}
      addFilter={jest.fn()}
      removeFilter={mockRemoveFilter}
    />,
  )

  fireEvent.click(getByText('G (2)'))

  await wait(() => expect(mockRemoveFilter).toHaveBeenCalled())
  expect(mockRemoveFilter.mock.calls[0]).toEqual(['recommendedKey'])
})

it('renders active options differently', () => {
  const { container: before } = renderUI(
    <SongFilter
      filters={testFilters}
      activeFilters={{}}
      addFilter={jest.fn()}
      removeFilter={jest.fn()}
    />,
  )

  const { container: after } = renderUI(
    <SongFilter
      filters={testFilters}
      activeFilters={{ recommendedKey: 'G' }}
      addFilter={jest.fn()}
      removeFilter={jest.fn()}
    />,
  )

  expect(before).toMatchDiffSnapshot(after)
})
