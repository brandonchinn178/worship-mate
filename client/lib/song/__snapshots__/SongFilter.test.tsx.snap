// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders active options differently 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -20,11 +20,11 @@
             (
            1
            )
          </div>
          <div
-           class=\\"SongFilter__FilterOption-sc-1nh6vn8-4 ciJMN\\"
+           class=\\"SongFilter__FilterOption-sc-1nh6vn8-4 GStQo\\"
          >
            G
             (
            2
            )"
`;
